compiler/src/dotty/tools/dotc/transform/BetaReduce.scala

```diff
@@ -38,51 +39,86 @@ class BetaReduce extends MiniPhase:
+  /** Rewrite an application
+   *
+   *    ((x1, ..., xn) => b)(e1, ..., en)
+   *
+   *  to
+   *
+   *    val/def x1 = e1; ...; val/def xn = en; b
+   *
+   *  where `def` is used for call-by-name parameters. However, we shortcut any NoPrefix
+   *  refs among the ei's directly without creating an intermediate binding.
+   *
+   *  Similarly, rewrites type applications
+   *
+   *    ([X1, ..., Xm] => (x1, ..., xn) => b).apply[T1, .., Tm](e1, ..., en)
+   *
+   *  to
+   *
+   *    type X1 = T1; ...; type Xm = Tm;val/def x1 = e1; ...; val/def xn = en; b
+   *
+   *  This beta-reduction preserves the integrity of `Inlined` tree nodes.
+   */
+  def apply(tree: Tree)(using Context): Tree =
+    val bindingsBuf = new ListBuffer[DefTree]
+    def recur(fn: Tree, argss: List[List[Tree]]): Option[Tree] = fn match
+      case Block((ddef : DefDef) :: Nil, closure: Closure) if ddef.symbol == closure.meth.symbol =>
+        Some(reduceApplication(ddef, argss, bindingsBuf))
+      case Block((TypeDef(_, template: Template)) :: Nil, Typed(Apply(Select(New(_), _), _), _)) if template.constr.rhs.isEmpty =>
+        template.body match
+          case (ddef: DefDef) :: Nil => Some(reduceApplication(ddef, argss, bindingsBuf))
+          case _ => None
+      case Block(stats, expr) if stats.forall(isPureBinding) =>
+        recur(expr, argss).map(cpy.Block(fn)(stats, _))
+      case Inlined(call, bindings, expr) if bindings.forall(isPureBinding) =>
+        recur(expr, argss).map(cpy.Inlined(fn)(call, bindings, _))
+      case Typed(expr, tpt) =>
+        recur(expr, argss)
+      case _ => None
+    tree match
+      case Apply(Select(fn, nme.apply), args) if defn.isFunctionType(fn.tpe) =>
+        recur(fn, List(args)) match
+          case Some(reduced) =>
+            seq(bindingsBuf.result(), reduced).withSpan(tree.span)
+          case None =>
+            tree
+      case Apply(TypeApply(Select(fn, nme.apply), targs), args) if fn.tpe.typeSymbol eq dotc.core.Symbols.defn.PolyFunctionClass =>
+        recur(fn, List(targs, args)) match
+          case Some(reduced) =>
+            seq(bindingsBuf.result(), reduced).withSpan(tree.span)
+          case None =>
+            tree
       case _ =>
-        original
-  end apply
-
-  /** Beta-reduces a call to `ddef` with arguments `args` */
-  def apply(ddef: DefDef, args: List[Tree])(using Context) =
-    val bindings = new ListBuffer[ValDef]()
-    val expansion1 = reduceApplication(ddef, args, bindings)
-    val bindings1 = bindings.result()
-    seq(bindings1, expansion1)
+        tree
 
   /** Beta-reduces a call to `ddef` with arguments `args` and registers new bindings */
-  def reduceApplication(ddef: DefDef, args: List[Tree], bindings: ListBuffer[ValDef])(using Context): Tree =
-    val vparams = ddef.termParamss.iterator.flatten.toList
-    assert(args.hasSameLengthAs(vparams))
+  def reduceApplication(ddef: DefDef, argss: List[List[Tree]], bindings: ListBuffer[DefTree])(using Context): Tree =
```

Comment:
It seems to me that `argss` is either a list of a list of terms, or a list consisting of a list of types followed by a list of terms. Is that correct? In that case, we should try to split type reduction and term reduction in two, so that we don't have to deal with lists of lists here. Two reasons:

 - List of lists is too general if all this reduces is maybe a type list followed by a term list.
 - We want to get away from the restriction that PolyFunctions need to have a term list following the type list. This PR makes that harder to change.

Reply:
> It seems to me that argss is either a list of a list of terms, or a list consisting of a list of types followed by a list of terms. Is that correct?

Yes. The idea for this representation is to be able generalize to SAM beta reduction. These would have an arbitrary number of parameter lists.
```scala
new T {
  def f[X](x: X)(y: Int): X = ...
}.f[Int](1)(2)
``` 
There is one case in the `InlinePatterns` where we reduce an `unapply`/`unapplySeq` in this way. This is special-cased and uses `reduceApplication` directly assuming we have the list of list of parameters. 

> We should try to split type reduction and term reduction in two, so that we don't have to deal with lists of lists here. 

That was my first approach. I opted against it because we would need to generate a new lambda after the type beta-reduction to then immediately beta-reduce that one. It seemed more complex and less performant.

> List of lists is too general if all this reduces is maybe a type list followed by a term list.

This list is intended to be the list of all arguments that the method receives.

> We want to get away from the restriction that PolyFunctions need to have a term list following the type list. This PR makes that harder to change.

I did consider this. The current implementation can handle methods with only type parameters. 

Furthermore, if we want to support interleaved type/term parameters (maybe extension method beta-reduction) we only need to change the line `val tparams = ddef.leadingTypeParams` to `val tparams = ddef.typeParamss`; then implement `typeParamss` in a similar way we implement`termParamss`.

Reply:
OK, that's convincing.